### PR TITLE
[MAINTENANCE] Removed some 30 type hint violations

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- JIRA: GREAT-1358/GREAT-1552
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Review by tannerbeam — APPROVED** [review]

File: great_expectations/expectations/metrics/metric_provider.py [hunk 1]
```diff
@@ -33,9 +33,9 @@ def wrapper(metric_fn: Callable):
         def inner_func(*args, **kwargs):
             return metric_fn(*args, **kwargs)
 
-        inner_func.metric_engine = engine
-        inner_func.metric_fn_type = MetricFunctionTypes(metric_fn_type)
-        inner_func.metric_definition_kwargs = kwargs
+        inner_func.metric_engine = engine  # type: ignore[attr-defined]
+        inner_func.metric_fn_type = MetricFunctionTypes(metric_fn_type)  # type: ignore[attr-defined]
+        inner_func.metric_definition_kwargs = kwargs  # type: ignore[attr-defined]
```
Comment (Kilo59) [inline great_expectations/expectations/metrics/metric_provider.py:38]: `non-blocking`
We could improve the typing information returned by this decorator with a `TypeVar` that preserves the original signature.
https://mypy.readthedocs.io/en/stable/generics.html?highlight=invariant#declaring-decorators

----------

Or alternatively, we could define a `Protocol` that represents these additional attributes that are being added and use that as the return type.

```python
from typing_extenstions import Protocol

class MyMetricProtocol(Protocol):
  @property
   def metric_engine(self) -> Type[ExecutionEngine]:
       ...

   @property
   def metric_definition_kwargs(self) -> dict]: # or TypedDict
       ...

   # etc.

...

    def wrapper(metric_fn: Callable):
        @wraps(metric_fn)
        def inner_func(*args, **kwargs) -> MyMetricProtocol:
```
Might even be able to improve on my example with a [Generic Protocol](https://mypy.readthedocs.io/en/stable/generics.html?highlight=invariant#generic-protocols) which might be able to preserve the details of the original callable. although I've never done this.


Non-blocking but if you don't do it as part of this PR can you create a follow-up ticket to explore these possibilities?

File: great_expectations/expectations/metrics/metric_provider.py [hunk 2]
```diff
@@ -96,14 +96,18 @@ class MetricProvider(metaclass=MetaMetricProvider):
...
 
     @classmethod
     def _register_metric_functions(cls) -> None:
         metric_name = getattr(cls, "metric_name", None)
         metric_domain_keys = cls.domain_keys
         metric_value_keys = cls.value_keys
 
+        if metric_name is None:
+            # No metric name has been defined
+            return
```
Comment (Kilo59) [inline great_expectations/expectations/metrics/metric_provider.py:109]:
```suggestion
        if not metric_name:
            # No metric name has been defined
            return
```

I expect we want an empty string to hit this branch as well?

File: great_expectations/expectations/metrics/metric_provider.py [hunk 3]
```diff
@@ -145,7 +150,9 @@ def _register_metric_functions(cls) -> None:
                     register_metric(
                         metric_name=declared_metric_name
                         + "."
-                        + metric_fn_type.metric_suffix,  # this will be a MetricPartial
+                        + cast(
+                            MetricPartialFunctionTypes, metric_fn_type
+                        ).metric_suffix,  # this will be a MetricPartial
```
Comment (Kilo59) [inline great_expectations/expectations/metrics/metric_provider.py:155]: Might be able to get rid of the cast with improvements to the `wrapper`

File: pyproject.toml [hunk 4]
```diff
@@ -154,9 +153,8 @@ exclude = [
     'render/renderer/v3/suite_profile_notebook_renderer\.py',  # 4
     'render/util\.py',  # 2
     'render/view/view\.py',  # 11
-    'rule_based_profiler/data_assistant_result/data_assistant_result\.py',  # 71
+    'rule_based_profiler/data_assistant_result/data_assistant_result\.py',  # 63
     'rule_based_profiler/data_assistant_result/plot_components\.py', # 12
-    'rule_based_profiler/data_assistant/data_assistant\.py',  # 15
```
Comment (Kilo59) [inline pyproject.toml:159]: 🚀

**Review by Kilo59 — APPROVED** [review]

File: great_expectations/expectations/metrics/metric_provider.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/metric_provider.py:109]: @Kilo59 Good idea (applying in next PR to be updated in a few minutes).  Thanks.

File: great_expectations/expectations/metrics/metric_provider.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/expectations/metrics/metric_provider.py:155]: @Kilo59 JIRA ticket created.  Thanks.

**Pull request merged**